src/lib/ast.rs

```diff
@@ -0,0 +1,617 @@
+        let next_index = self.nodes.len();
+        let node = Node::new(data, ty, indent);
+        self.nodes.push(node);
+        next_index
+    }
+
+    /// Return `true` if the arena is empty, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /// Make one node the child of another.
+    /// Panic if either `NodeId` is not found.
+    pub fn make_child_of(&mut self, child: NodeId, parent: NodeId) {
+        if child >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", child);
+        }
+        if parent >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", parent);
+        }
+        self[child].parent = Some(parent);
+        match self[parent].first_child {
+            None => self[parent].first_child = Some(child),
+            Some(_) => (),
+        };
+        match self[parent].last_child {
+            None => self[parent].last_child = Some(child),
+            Some(id) => {
+                self[parent].last_child = Some(child);
+                self[id].next_sibling = Some(child);
+                self[child].previous_sibling = Some(id);
+            }
+        };
+    }
+
+    // Return a vector of edges between nodes in this arena.
+    // Each edge pair `(n1, n2)` denotes an edge between nodes `n1` and `n2`,
+    // both numbers being `NodeId.index` indices within this arena.
+    fn get_edges(&self) -> Vec<EdgeId> {
+        let mut edges: Vec<EdgeId> = vec![];
+        for index in 0..self.nodes.len() {
+            let node = &self.nodes[index];
+            match node.parent {
+                None => (),
+                Some(par) => edges.push((par, index)),
+            }
+        }
+        edges
+    }
+}
+
+impl Arena<String, String> {
+    /// Create a graphviz representation of this arena and write to buffer.
+    pub fn render_dotgraph<W: Write>(&self, buffer: &mut W) -> Result<(), io::Error> {
+        dot::render(self, buffer)
+    }
+}
+
+impl<'a> dot::Labeller<'a, NodeId, EdgeId> for Arena<String, String> {
+    // The graph and node ids here must be non-empty strings made up of
+    // alphanumeric or underscore characters, not beginning with a digit (i.e.
+    // the regular expression [a-zA-Z_][a-zA-Z_0-9]*). This means we cannot
+    // label nodes with arithmetic operators, etc.
+
+    fn graph_id(&self) -> dot::Id {
+        dot::Id::new("AST").unwrap()
+    }
+
+    fn node_id(&self, id: &NodeId) -> dot::Id {
+        // Node ids must be unique.
+        dot::Id::new(format!("N{}", id)).unwrap()
+    }
+
+    fn node_label(&self, id: &NodeId) -> dot::LabelText {
+        let mut label = String::new();
+        label.push_str(self.nodes[*id].ty.as_str());
+        // Only add node.data to the label if it is alphanumeric.
+        let mut is_data_alphanum = true;
+        if self.nodes[*id].data.len() > 0 {
+            for character in self.nodes[*id].data.chars() {
```

Comment:
It might be more idiomatic to do something like:
```rust
if self.nodes[*id].data.chars().iter().find(|c| c.is_alphanumeric()).is_some() {
    label.push(...);
    ...
}
```